Add '--no-onboarding' flag to 'generateuserdata'

jonboiser opened the pull request:
To support automated testing of upgrades #2200 , I wanted a way to provision a kolibri server quickly and run tests without having to go through the onboarding flow.

This PR:

1. Moves a lot of the code in `Command.handle` to an external utilities module to clean up the code.
2. Adds a new flag `--no-onboarding`, which "provisions" the device and adds admin users to each facility. So after running `kolibri manage generateuserdata --no-onboarding`, one can start the server and not have to go through onboarding.

rtibbles commented on kolibri/logger/management/commands/userdata_utils.py:
```diff
@@ -0,0 +1,327 @@
+from le_utils.constants import content_kinds
+
+
+def provision_device():
```
Can consolidate this with this: https://github.com/learningequality/kolibri/blob/release-v0.6.x/kolibri/auth/test/helpers.py#L17

rtibbles commented on kolibri/logger/management/commands/userdata_utils.py:
```diff
@@ -0,0 +1,327 @@
+from __future__ import absolute_import, print_function, unicode_literals
```
Maybe let's move this into `logger/utils` - so as to avoid having non-management commands in the commands folder.

rtibbles commented on kolibri/logger/management/commands/userdata_utils.py:
```diff
@@ -0,0 +1,327 @@
+    device_settings.save()
+
+
+def add_superuser_to_facility(facility):
```
Ditto here: https://github.com/learningequality/kolibri/blob/release-v0.6.x/kolibri/auth/test/helpers.py#L9

rtibbles commented on kolibri/logger/management/commands/userdata_utils.py:
```diff
@@ -0,0 +1,327 @@
+    # Full names - because we should never assume that users have names like us
+    user_data_name_fields = ["GivenName", "MiddleInitial", "Surname"]
+
+    n_on_device = HierarchyRelationsFilter(FacilityUser.objects.all()).filter_by_hierarchy(
```
This is a slightly misleading variable name, as it is the number of users in the classroom, not on the device in total.

rtibbles commented on kolibri/logger/management/commands/userdata_utils.py:
```diff
@@ -0,0 +1,327 @@
+            )
+            # Set a dummy password so that if we want to login as this learner later, we can.
+            user.set_password('password')
+            # Add the user to the current classroom
```
Residual bug from my code - we should call `user.save()` after setting the password, otherwise this password change is not saved.

benjaoming replied:
https://github.com/learningequality/kolibri/issues/2265 :P

rtibbles commented on kolibri/logger/management/commands/userdata_utils.py:
```diff
@@ -0,0 +1,327 @@
+    )[0:n_users]
+
+
+def add_channel_activity_for_user(**options): # noqa: max-complexity=16
```
Is the `noqa` flag still necessary here? (possibly, but with your slight refactor, it might have dipped).

rtibbles commented on kolibri/logger/management/commands/generateuserdata.py:
```diff
@@ -30,307 +23,61 @@ def add_arguments(self, parser):
         # on the device than specified
         parser.add_argument('--facilities', type=int, default=1, dest="facilities", help="Number of facilities")
+        parser.add_argument('--no-onboarding', action='store_true', dest="no_onboarding", help="Automatically create superusers and skip onboarding")
 
```
If you want to add a `now` argument, can add one here too? I imagine you can set the default to a callable `timezone.now`.

jonboiser replied:
What should the type of that be? String? How would a user enter a timestamp on the command line?

rtibbles replied:
We would probably have to decide on a standard strftime compatible string. Let's leave it for now, we can add it if there is a demand for it!

benjaoming commented on kolibri/logger/management/commands/generateuserdata.py:
```diff
@@ -30,307 +23,61 @@ def add_arguments(self, parser):
-                                            # Otherwise only let students get odd indexed questions right,
-                                            # ensuring they will always have a mastery breaking sequence
-                                            # as zero based indexing means their first attempt will always be wrong!
-                                            correct = j % 2 == 1
```
Amazing, didn't this code fail the linting? McCabe complexity level seems extremely high!!

rtibbles replied:
That's why there was a `#  noqa` at the top. I know, because I copied it from where you did the same thing somewhere else :)

benjaoming replied:
Hahaha, *blushes*

jonboiser commented:
@rtibbles 

1. Re-used stuff from kolibri.auth.test.helpers in the command
1. Moved userdata_helpers.py to logger.utils
1. Addressed line comments

rtibbles commented on kolibri/logger/utils/user_data.py:
```diff
@@ -0,0 +1,306 @@
+    )[0:n_users]
+
+
+def add_channel_activity_for_user(**options): # noqa: max-complexity=16
```
Might be a little cleaner to use args rather than kwargs here, but fine for now.

indirectlylit replied:
worth making a low-priority 'tech update' issue?
